core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala

```diff
@@ -1051,6 +1049,15 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
     val jobAttemptId = newTaskAttemptID(jobtrackerID, stageId, isMap = true, 0, 0)
     val jobTaskContext = newTaskAttemptContext(wrappedConf.value, jobAttemptId)
     val jobCommitter = jobFormat.getOutputCommitter(jobTaskContext)
+
+    // If speculation is enabled, we only allow FileOutputCommitter.
+    val speculationEnabled = self.conf.getBoolean("spark.speculation", false)
+    if (speculationEnabled && jobCommitter.getClass !=
+      classOf[org.apache.hadoop.mapreduce.lib.output.FileOutputCommitter]) {
+      throw new SparkException(s"Cannot use ${jobCommitter.getClass} as output committer when" +
+        "speculation is enabled.")
+    }
+
```

Comment:
I agree this is a too strong limitation, but my concern is that: there is possibility that a user-defined output formatter provides a direct output committer which we should reject. However, these is no simple way to see if a output committer is direct or not, so I just restrict it to `FileOutputCommitter` only here.